### PR TITLE
feat(trajectory_follower): integrate latlon controller

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

Integate lateral and longitudinal controllers into single node.
- Each controller(pid, mpc, pure_pursuit) inherits the base controller. 
-  Control data can be shared (synced) between lateral and longitudinal. 
   - Share steer convergence information to the longitudinal controller. This allows the steer to move while stopped.
image

## Related links

launcher https://github.com/tier4/autoware_launch/pull/366
## Tests performed
I checked that all test codes pass.
`colcon test --event-handlers console_cohesion+ --packages-select trajectory_follower_nodes` 

Also checked there is no degradation with scenario tests.
These are TIER IV internal test results but I share them for your imformation.
-  pilot-auto.xx1 [beta/v0.31.2.0-test](https://github.com/tier4/pilot-auto.xx1/tree/beta/v0.31.2.0-test)
   -  279/292 passed
      - https://evaluation.tier4.jp/evaluation/reports/3e18f971-c932-5698-839f-1f1077d62eab?project_id=prd_jt 
- pilot-auto.xx1 [tmp/latlon_on_v0.31.2.0-test](https://github.com/tier4/pilot-auto.xx1/tree/tmp/latlon_on_v0.31.2.0-test) (with this PR after @TakaHoribe review)
   - 280/292 passed (with converged `false`)
      - https://evaluation.tier4.jp/evaluation/reports/6a99c070-32eb-5af5-ab7c-6b1408c3d88e?project_id=prd_jt
   - 285/292 passed (with converged `true`)
      - https://evaluation.tier4.jp/evaluation/reports/d7c47e49-c2c9-59b2-8c83-4e7d8a8cd570?project_id=prd_jt

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: control/trajectory_follower/include/trajectory_follower/lateral_controller_base.hpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+{
+  autoware_auto_control_msgs::msg::AckermannLateralCommand control_cmd;
+  LateralSyncData sync_data;
+};
+
+class LateralControllerBase
+{
+public:
+  virtual LateralOutput run() = 0;
+  void sync(LongitudinalSyncData longitudinal_sync_data)
```
Comment (takayuki5168) [inline control/trajectory_follower/include/trajectory_follower/lateral_controller_base.hpp:41]: const reference

File: control/trajectory_follower/include/trajectory_follower/longitudinal_controller_base.hpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+struct LongitudinalOutput
+{
+  autoware_auto_control_msgs::msg::LongitudinalCommand control_cmd;
+  LongitudinalSyncData sync_data;
+};
+class LongitudinalControllerBase
+{
+public:
+  virtual LongitudinalOutput run() = 0;
+  void sync(LateralSyncData lateral_sync_data) { lateral_sync_data_ = lateral_sync_data; };
```
Comment (takayuki5168) [inline control/trajectory_follower/include/trajectory_follower/longitudinal_controller_base.hpp:40]: const reference

File: control/trajectory_follower_nodes/CMakeLists.txt [hunk 3]
```diff
@@ -46,6 +46,18 @@ rclcpp_components_register_node(${LATLON_MUXER_NODE}
   EXECUTABLE ${LATLON_MUXER_NODE}_exe
 )
 
+set(CONTROLLER_NODE controller_node)
+ament_auto_add_library(${CONTROLLER_NODE} SHARED
+include/trajectory_follower_nodes/controller_node.hpp
```
Comment (takayuki5168) [inline control/trajectory_follower_nodes/CMakeLists.txt:51]: wrong indent

File: control/trajectory_follower/include/trajectory_follower/longitudinal_controller_base.hpp (hunk 2 above)
Comment (kosuke55) [reply control/trajectory_follower/include/trajectory_follower/longitudinal_controller_base.hpp]: https://github.com/autowarefoundation/autoware.universe/pull/901/commits/7970d4f1cf52700dbe7f21cc92f93881669fa444

File: control/trajectory_follower/include/trajectory_follower/lateral_controller_base.hpp (hunk 1 above)
Comment (kosuke55) [reply control/trajectory_follower/include/trajectory_follower/lateral_controller_base.hpp]: https://github.com/autowarefoundation/autoware.universe/pull/901/commits/7970d4f1cf52700dbe7f21cc92f93881669fa444

**Comment by kosuke55** [comment]
PRs that need to be captured
https://github.com/autowarefoundation/autoware.universe/pull/1063
https://github.com/autowarefoundation/autoware.universe/pull/1103

File: control/trajectory_follower/src/pid_longitudinal_controller.cpp [hunk 4]
```diff
@@ -448,14 +464,40 @@ LongitudinalController::Motion LongitudinalController::calcEmergencyCtrlCmd(
...
+      tier4_autoware_utils::calcDistance2d(
+        trajectory.points.back().pose, m_trajectory_ptr->points.back().pose) >
+      p.stopped_state_new_traj_end_dist) {
+      return true;
+    }
+  }
+  return false;
+}
+
+PidLongitudinalController::ControlState PidLongitudinalController::updateControlState(
```
Comment (kosuke55) [inline control/trajectory_follower/src/pid_longitudinal_controller.cpp:500]: It takes time to respond to a new trajectory, so the steer is initially in a state of convergence.

https://user-images.githubusercontent.com/39142679/175781079-a93f6175-db14-4c1e-8a8c-0668fb612c82.mp4

Comment (kosuke55) [reply control/trajectory_follower/src/pid_longitudinal_controller.cpp]: ( rename checkNewTrajectory -> isNewTrajectory )

File: control/trajectory_follower_nodes/src/controller_node.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  throw std::domain_error("[LateralController] undesired algorithm is selected.");
+  return LateralControllerMode::INVALID;
+}
+
+Controller::LongitudinalControllerMode Controller::getLongitudinalControllerMode(
+  const std::string & controller_mode) const
+{
+  if (controller_mode == "pid") return LongitudinalControllerMode::PID;
+
+  throw std::domain_error("[LongitudinalController] undesired algorithm is selected.");
```
Comment (takayuki5168) [inline control/trajectory_follower_nodes/src/controller_node.cpp:106]: This should be removed.

File: control/trajectory_follower_nodes/src/controller_node.cpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  }
+}
+
+Controller::LateralControllerMode Controller::getLateralControllerMode(
+  const std::string & controller_mode) const
+{
+  if (controller_mode == "mpc_follower") return LateralControllerMode::MPC;
+  if (controller_mode == "pure_pursuit") return LateralControllerMode::PURE_PURSUIT;
+
+  throw std::domain_error("[LateralController] undesired algorithm is selected.");
```
Comment (takayuki5168) [inline control/trajectory_follower_nodes/src/controller_node.cpp:97]: This should be removed.

File: control/trajectory_follower_nodes/src/controller_node.cpp (hunk 5 above)
Comment (kosuke55) [reply control/trajectory_follower_nodes/src/controller_node.cpp]: remove in https://github.com/autowarefoundation/autoware.universe/pull/901/commits/2d595771f5f018b62eca310dc57351dcd85bb7de

File: control/trajectory_follower_nodes/src/controller_node.cpp (hunk 6 above)
Comment (kosuke55) [reply control/trajectory_follower_nodes/src/controller_node.cpp]: remove in https://github.com/autowarefoundation/autoware.universe/pull/901/commits/2d595771f5f018b62eca310dc57351dcd85bb7de

**Comment by kosuke55** [comment]
@TakaHoribe thank you for your review. I checked no degradation with scenario tests after the review.
279/292 passed 
https://evaluation.tier4.jp/evaluation/reports/f64556b2-eac2-5607-8217-eab497344038?project_id=prd_jt

File: control/trajectory_follower_nodes/param/lateral_controller_defaults.param.yaml [hunk 7]
```diff
@@ -56,7 +56,9 @@
     stop_state_entry_ego_speed: 0.001
     stop_state_entry_target_speed: 0.001
     converged_steer_rad: 0.1
-    check_steer_converged_for_stopped_state: true
+    keep_steer_control_until_converged: true
```
Comment (kosuke55) [inline control/trajectory_follower_nodes/param/lateral_controller_defaults.param.yaml:59]: better to set false by default

File: launch/tier4_control_launch/config/trajectory_follower/lateral_controller.param.yaml [hunk 8]
```diff
@@ -57,7 +57,9 @@
     stop_state_entry_ego_speed: 0.001
     stop_state_entry_target_speed: 0.001
     converged_steer_rad: 0.1
-    check_steer_converged_for_stopped_state: true
+    keep_steer_control_until_converged: true
```
Comment (kosuke55) [inline launch/tier4_control_launch/config/trajectory_follower/lateral_controller.param.yaml:60]: better to set false by default

File: launch/tier4_control_launch/config/trajectory_follower/longitudinal_controller.param.yaml [hunk 9]
```diff
@@ -1,11 +1,11 @@
 /**:
   ros__parameters:
-    longitudinal_ctrl_period: 0.03  # control period [s]
     delay_compensation_time: 0.17
 
     enable_smooth_stop: true
     enable_overshoot_emergency: true
     enable_slope_compensation: false
+    enable_keep_stopped_until_steer_convergence: true
```
Comment (kosuke55) [inline launch/tier4_control_launch/config/trajectory_follower/longitudinal_controller.param.yaml:8]: better to set false by default

**Comment by kosuke55** [comment]
need to capture https://github.com/autowarefoundation/autoware.universe/pull/1201

File: launch/tier4_control_launch/config/trajectory_follower/longitudinal_controller.param.yaml (hunk 9 above)
Comment (kosuke55) [reply launch/tier4_control_launch/config/trajectory_follower/longitudinal_controller.param.yaml]: [b51a021](https://github.com/autowarefoundation/autoware.universe/pull/901/commits/b51a021429fcd6bd41642c52866f154b4797bdc6)

File: control/trajectory_follower_nodes/param/longitudinal_controller_defaults.param.yaml [hunk 10]
```diff
@@ -1,11 +1,11 @@
 /**:
   ros__parameters:
-    longitudinal_ctrl_period: 0.03  # control period [s]
     delay_compensation_time: 0.17
 
     enable_smooth_stop: true
     enable_overshoot_emergency: true
     enable_slope_compensation: false
+    enable_keep_stopped_until_steer_convergence: true
```
[the remark this answers is not in the document]
Comment (kosuke55) [reply control/trajectory_follower_nodes/param/longitudinal_controller_defaults.param.yaml]: [b51a021](https://github.com/autowarefoundation/autoware.universe/pull/901/commits/b51a021429fcd6bd41642c52866f154b4797bdc6)

File: launch/tier4_control_launch/config/trajectory_follower/lateral_controller.param.yaml (hunk 8 above)
Comment (kosuke55) [reply launch/tier4_control_launch/config/trajectory_follower/lateral_controller.param.yaml]: [b51a021](https://github.com/autowarefoundation/autoware.universe/pull/901/commits/b51a021429fcd6bd41642c52866f154b4797bdc6)

File: control/trajectory_follower_nodes/param/lateral_controller_defaults.param.yaml (hunk 7 above)
Comment (kosuke55) [reply control/trajectory_follower_nodes/param/lateral_controller_defaults.param.yaml]: [b51a021](https://github.com/autowarefoundation/autoware.universe/pull/901/commits/b51a021429fcd6bd41642c52866f154b4797bdc6)

**Review by TakaHoribe — APPROVED** [review]
LGTM

**Comment by TakaHoribe** [comment]
Please update readme. It could be a separate PR.

**Pull request merged**